### DISCUSSION TITLE
Potential ReDoS Vulnerability or Inefficient Regular Expression in Project: Need for Assessment and Mitigation

### DISCUSSION THREAD

**Discussion by SuperMaxine** [sxyazi/yazi#457] [1. Q&A]
Hello,

I am writing to report a potential Regular Expression Denial of Service (ReDoS) vulnerability or Inefficient Regular Expression in the project. This issue arises when specially crafted input strings are used in the context of distributed, high-volume requests, potentially leading to a denial-of-service attack.

**Location of Issue:**

The vulnerability is related to a regular expression used in the following validation file, which may result in significantly prolonged execution times under certain conditions.

https://github.com/sxyazi/yazi/blob/d6ca524c53053d4670bdd920f40d2515c1bc6761/yazi-shared/src/fs/path.rs#L18

**PoC Files and Comparisons:**

[PoC.zip](https://github.com/sxyazi/yazi/files/13778119/PoC.zip)

To evaluate the performance of this inefficient regular expression matching with varying input contents, the following commands can be executed within the PoC folder:

```bash
cargo build --release # Build the PoC executable file
time ./target/release/redos AttackString10MB.txt
time ./target/release/redos RandomString10MB.txt
time ./target/release/redos AttackString1MB.txt
time ./target/release/redos RandomString1MB.txt
```

These commands measure the time taken for the regular expression to match different types of strings. The results on my machine are as follows: For a 10MB attack string, the processing time exceeded 10 minutes. In contrast, a 10MB random string took only 0.016 seconds. A 1MB attack string took 6.099 seconds. By comparison, a 1MB random string took merely 0.005 seconds.

**Proposed Solution:**

A possible mitigation strategy could include limiting the input length to prevent excessive processing times.

**Additional Considerations:**

Historically, it was believed that using regex engines with non-backtracking implementations (such as those in Rust or Go) would not lead to ReDoS vulnerabilities. However, recent studies have shown that this is not always the case. I recommend an assessment of how this issue might impact this project.

Thank you for your attention to this matter. Your evaluation and response to this potential security concern would be greatly appreciated.

Best regards,

**Comment by sxyazi** [comment]
Thank you for the report.

AFAIK, paths in Linux/macOS/Windows all have length limitations, [generally quite low](https://www.ibm.com/docs/en/spectrum-protect/8.1.9?topic=parameters-file-specification-syntax). Yazi's path originates from the user's file system, and I don't believe there would be a file name with a length of 1MB; the file system wouldn't allow it.
